### PR TITLE
feat: add length prediction module with test and distribution modeling

### PR DESCRIPTION
**Pull request by ImYangYun**
In order to generate future motion sequences of variable length, we propose predicting the target length from past motion and sign input. This PR uses the distribution module into the SignWritingToPoseDiffusion model, enabling explicit prediction of the future pose sequence length as a Gaussian distribution. Key changes include:

- distribution.py: Implements DiagonalGaussianDistribution for modeling the predicted mean and log-variance of future sequence lengths. Provides sampling and negative log-likelihood methods.
- models.py: Adds length_predictor, which takes the global latent (aggregated from timestep, signwriting, motion encoders) and predicts a Gaussian distribution over target length.
- test_length_prediction.py: Adds a micro training loop to validate the length predictor’s overfit capability on small real data batches. It allows training and evaluating the length predictor with visualized loss metrics (NLL, MSE, MAE).
- data_loader.py: Refactors target pose cropping logic to support dynamic length extraction and expose length_target as part of each sample.

**Current Issues:** 
Length prediction accuracy is inconsistent: The relative error threshold (< 0.7) often fails on shorter target sequences (e.g., length 1–5), where small absolute deviations lead to disproportionately high relative errors. This leads to test failures even when predictions are qualitatively reasonable.

**Visual Results:**
See attached loss curves for convergence behavior over 500 steps of micro-training.
length_nll_loss

length_prediction_scatter

### CODE REVIEW

File: signwriting_animation/diffusion/core/distribution.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import torch
+from torch import nn
+
+class DiagonalGaussianDistribution:
```
Comment (AmitMY) [inline signwriting_animation/diffusion/core/distribution.py:4]: what makes it diagonal?
Simple sampling model: https://github.com/sign-language-processing/transcription/blob/main/text_to_pose/diffusion/src/model/distribution.py

File: signwriting_animation/diffusion/tests/test_length_prediction.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    print("Mean NLL:            ", round(float(nll.mean()), 4))
+
+    relative_error = abs_diff / target_lengths.clamp(min=1.0)
+    max_relative_error = 0.7
+
+    if not torch.all(relative_error < max_relative_error):
+        print("Relative error exceeded for some samples!")
+        print("Relative errors:", relative_error.tolist())
+        print("Target lengths:", target_lengths.tolist())
+        print("Predicted means:", pred_lengths.tolist())
```
Comment (AmitMY) [inline signwriting_animation/diffusion/tests/test_length_prediction.py:115]: this is too explicit - you can use the trainer, same as we do in training. no need to ever call `model.global_norm` etc - or calculating the loss.

The trainer can have callbacks, to log the loss etc. You can give the test to claude or chatgpt to improve it.

File: signwriting_animation/diffusion/tests/test_length_prediction.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+from pose_format.torch.masked.collator import zero_pad_collator
+import matplotlib.pyplot as plt
+import os
+
+@pytest.mark.parametrize("batch_size", [4])
+def test_length_prediction_on_real_data(batch_size):
+    data_dir = "/scratch/yayun/pose_data"
+    csv_path = "/data/yayun/signwriting-animation/data.csv"
+
+    dataset = DynamicPosePredictionDataset(
```
Comment (AmitMY) [inline signwriting_animation/diffusion/tests/test_length_prediction.py:15]: you could create a "mock" dataset in order to test just this rather than the entire process

File: signwriting_animation/diffusion/core/models.py [hunk 4]
```diff
@@ -126,7 +128,11 @@ def forward(self,
         xseq = self.sequence_pos_encoder(xseq)
         output = self.seqEncoder(xseq)[-num_frames:]
         output = self.pose_projection(output)
-        return output
+        global_latent = self.global_norm(xseq.mean(0))
+        print("Global latent stats: min", global_latent.min().item(), "max", global_latent.max().item()) 
+
+        length_dist = self.length_predictor(global_latent)
+        return output, length_dist
```
Comment (shaunirwin) [inline signwriting_animation/diffusion/core/models.py:135]: Would be good to update the docstring of the `forward` function to show the type, description and shape of the additional output variable

File: signwriting_animation/diffusion/tests/test_length_prediction.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+from signwriting_animation.data.data_loader import DynamicPosePredictionDataset, get_num_workers
+from signwriting_animation.diffusion.core.models import SignWritingToPoseDiffusion
+from pose_format.torch.masked.collator import zero_pad_collator
+import matplotlib.pyplot as plt
+import os
+
+@pytest.mark.parametrize("batch_size", [4])
+def test_length_prediction_on_real_data(batch_size):
+    data_dir = "/scratch/yayun/pose_data"
+    csv_path = "/data/yayun/signwriting-animation/data.csv"
```
[the remark this answers is not in the document]
Comment (shaunirwin) [reply signwriting_animation/diffusion/tests/test_length_prediction.py:13]: could use the existing test dataset in `signwriting_animation/diffusion/tests/test_dataset.py`

File: signwriting_animation/diffusion/core/distribution.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+        self.fc_mu = nn.Linear(input_size, 1)
+        self.fc_var = nn.Linear(input_size, 1)
+
+    def forward(self, x: torch.Tensor):
+        mu = self.fc_mu(x)
+        log_var = self.fc_var(x)
+        std = torch.exp(0.5 * log_var)
+        q = torch.distributions.Normal(mu, std)
+        return q
```
Comment (AmitMY) [inline signwriting_animation/diffusion/core/distribution.py:17]: missing new line at end of file

**Review by AmitMY — CHANGES_REQUESTED** [review]
Great changes:
1. please add the new test plots in a comment
2. please make sure it passes lint

**Comment by ImYangYun** [comment]
> Regarding the pylint warnings for number of args, one option is to pass in related arguments as a single python dataclass. e.g:
> 
> ```
> from dataclasses import dataclass
> from typing import Literal
> 
> @dataclass
> class DatasetConfig:
>     data_dir: str
>     csv_path: str
>     num_past_frames: int = 40
>     num_future_frames: int = 20
>     split: Literal['train', 'test', 'dev'] = 'train'
> 
> class DynamicPosePredictionDataset(Dataset):
>     def __init__(self,  
>          config: DatasetConfig,
>          clip_model_name: str = "openai/clip-vit-base-patch32",
>          with_metadata: bool = True):
>         ...
> ```
> 
> I'm not sure if the `with_metadata` is super necessary: maybe just always return metadata? Unless there are reasons that you can think of where we might not want to.

Thanks! I'll take a look at it!

**Review by AmitMY — APPROVED** [review]
Both linting and tests fail. please fix, so i can merge
